src/stratigraphy/data_extractor/data_extractor.py

```diff
@@ -0,0 +1,143 @@
+from stratigraphy.util.line import TextLine
+from stratigraphy.util.util import read_params
+
+logger = logging.getLogger(__name__)
+
+
+class DataExtractor(ABC):
+    """Abstract class for data extraction from stratigraphy data files.
+
+    This class defines the interface for extracting data from stratigraphy data files. Subclasses must implement the
+    extract_data method to define the data extraction logic.
+    """
+
+    doc: fitz.Document = None
+    feature_keys: list[str] = None
+    feature_name = None
+
+    def __init__(self, document: fitz.Document, feature_name: str = "coordinate"):
+        """Initializes the CoordinateExtractor object.
+
+        Args:
+            document (fitz.Document): A PDF document.
+            feature_name (str, optional): The name of the feature to extract. Defaults to "coordinate".
+        """
+        self.doc = document
+        self.feature_name = feature_name
+        self.feature_keys = read_params("matching_params.yml")[f"{self.feature_name}_keys"]
+
+    def find_feature_key(self, lines: list[TextLine], allowed_errors: int = 3) -> TextLine | None:  # noqa: E501
+        """Finds the location of a coordinate key in a string of text.
+
+        This is useful to reduce the text within which the coordinates are searched. If the text is too large
+        false positive (found coordinates that are no coordinates) are more likely.
+
+        The function allows for a certain number of errors in the key. Errors are defined as insertions, deletions
+        or substitutions of characters (i.e. Levenshtein distance). For more information of how errors are defined see
+        https://github.com/mrabarnett/mrab-regex?tab=readme-ov-file#approximate-fuzzy-matching-hg-issue-12-hg-issue-41-hg-issue-109.
+
+
+        Args:
+            lines (list[TextLine]): Arbitrary text lines to search in.
+            allowed_errors (int, optional): The maximum number of errors (Levenshtein distance) to consider a key
+                                            contained in text. Defaults to 3 (guestimation; no optimisation done yet).
+
+        Returns:
+            TextLine | None: The line of the coordinate key found in the text.
+        """
+        matches = []
+        for key in self.feature_keys:
+            if len(key) < 5:
+                # if the key is very short, do an exact match
+                pattern = regex.compile(r"\b" + key + r"\b", flags=regex.IGNORECASE)
+            else:
+                pattern = regex.compile(r"\b(" + key + "){e<" + str(allowed_errors) + r"}\b", flags=regex.IGNORECASE)
+
+            for line in lines:
+                match = pattern.search(line.text)
+                if match:
+                    matches.append((line, sum(match.fuzzy_counts)))
+
+        # if no match was found, return None
+        if len(matches) == 0:
+            return None
+
+        # Remove duplicates
+        matches = list(dict.fromkeys(matches))
```

Comment:
For groundwater extraction, I implemented this differently in https://github.com/swisstopo/swissgeol-boreholes-dataextraction/pull/72/commits/ee46a8267fdff082660ad5f7a0a107fe70e51ea9 . I believe that that that implementation might also be better for the other attributes.

Maybe this is also what is causing the merge conflict that Github is currently reporting?

Reply:
Thank you for the detailed look into the PR. Yes, I do believe that I still need to merge with the latest main branch which will add your changes. Currently, there is also still the n-best logic in there that we removed.